rustls-postquantum/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+        let combined_share = CombinedShare::combine(&x25519.pub_key, kyber_share);
+
+        Ok(CompletedKeyExchange {
+            group: self.name(),
+            pub_key: combined_share.0,
+            secret: SharedSecret::from(&combined_secret.0[..]),
+        })
+    }
+
+    fn name(&self) -> NamedGroup {
+        NAMED_GROUP
+    }
+}
+
+struct Active {
+    x25519: Box<dyn ActiveKeyExchange>,
+    decap_key: Box<kem::DecapsulationKey<AlgorithmId>>,
+    combined_pub_key: Vec<u8>,
+}
+
+impl ActiveKeyExchange for Active {
+    fn complete(self: Box<Self>, peer_pub_key: &[u8]) -> Result<SharedSecret, Error> {
+        let ciphertext = match ReceivedCiphertext::new(peer_pub_key) {
+            Some(ct) => ct,
+            None => {
+                return Err(INVALID_KEY_SHARE);
+            }
+        };
+
+        let combined = CombinedSecret::combine(
+            self.x25519
+                .complete(ciphertext.x25519)?,
+            self.decap_key
+                .decapsulate(ciphertext.kyber.into())
+                .map_err(|_| INVALID_KEY_SHARE)?,
+        );
+        Ok(SharedSecret::from(&combined.0[..]))
+    }
+
+    fn pub_key(&self) -> &[u8] {
+        &self.combined_pub_key
+    }
+
+    fn group(&self) -> NamedGroup {
+        NAMED_GROUP
+    }
+}
+
+fn kyber768_r3() -> &'static kem::Algorithm<AlgorithmId> {
+    get_algorithm(AlgorithmId::Kyber768_R3).expect("Kyber768_R3 not available")
+}
+
+struct ReceivedShare<'a> {
+    x25519: &'a [u8],
+    kyber: &'a [u8],
+}
+
+impl<'a> ReceivedShare<'a> {
+    fn new(buf: &'a [u8]) -> Option<ReceivedShare<'a>> {
+        if buf.len() != COMBINED_PUBKEY_LEN {
+            return None;
+        }
+
+        let (x25519, kyber) = buf.split_at(X25519_LEN);
+        Some(ReceivedShare { x25519, kyber })
+    }
+}
+
+struct ReceivedCiphertext<'a> {
+    x25519: &'a [u8],
+    kyber: &'a [u8],
+}
+
+impl<'a> ReceivedCiphertext<'a> {
+    fn new(buf: &'a [u8]) -> Option<ReceivedCiphertext<'a>> {
+        if buf.len() != COMBINED_CIPHERTEXT_LEN {
+            return None;
+        }
+
+        let (x25519, kyber) = buf.split_at(X25519_LEN);
+        Some(ReceivedCiphertext { x25519, kyber })
+    }
+}
+
+struct CombinedSecret([u8; COMBINED_SHARED_SECRET_LEN]);
+
+impl CombinedSecret {
+    fn combine(x25519: SharedSecret, kyber: kem::SharedSecret) -> Self {
+        let mut out = CombinedSecret([0u8; COMBINED_SHARED_SECRET_LEN]);
+        out.0[..X25519_LEN].copy_from_slice(x25519.secret_bytes());
+        out.0[X25519_LEN..].copy_from_slice(kyber.as_ref());
+        out
+    }
+}
+
+struct CombinedShare(Vec<u8>);
+
+impl CombinedShare {
+    fn combine(x25519: &[u8], kyber: kem::Ciphertext) -> Self {
+        let mut out = CombinedShare(vec![0u8; COMBINED_CIPHERTEXT_LEN]);
+        out.0[..X25519_LEN].copy_from_slice(x25519);
+        out.0[X25519_LEN..].copy_from_slice(kyber.as_ref());
+        out
+    }
+}
+
+const NAMED_GROUP: NamedGroup = NamedGroup::Unknown(0x6399);
```

Comment:
I think it's unrelated to this branch, but I wonder if the `Unknown` variant could have a more descriptive name. In this context it _is_ known, just not to Rustls, but to the extension crypto provider :thinking: 

Maybe `NamedGroup::Extension`? I'll give it more thought. Just rubber ducking here.

Reply:
`Other` perhaps?

Reply:
I like it, I'll put up a separate rename PR so we can resolve this thread.

Reply:
I think this would be a breaking change -- maybe we can leave it for later?

Reply:
(Or alternatively we could branch for 0.23 maintenance.)